### PR TITLE
[SPARK-15323][SPARK-14463][SQL] Fix reading of partitioned format=text datasets

### PR DESCRIPTION
**Pull request by jurriaan**
https://issues.apache.org/jira/browse/SPARK-15323

I was using partitioned text datasets in Spark 1.6.1 but it broke in Spark 2.0.0.

It would be logical if you could also write those,
but not entirely sure how to solve this with the new DataSet implementation.

Also it doesn't work using `sqlContext.read.text`, since that method returns a `DataSet[String]`.
See https://issues.apache.org/jira/browse/SPARK-14463 for that issue.

### CODE REVIEW

**Comment by marmbrus** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
(I think it might need a JIRA because it seems changing existing behaviour)

**Comment by jurriaan** [comment]
@HyukjinKwon It's related to https://issues.apache.org/jira/browse/SPARK-14463. Or should I create a new JIRA?

And how is this changing existing behaviour? It was working perfectly fine in Spark 1.6.1.

**Comment by HyukjinKwon** [comment]
Oh, I just meant it changes codes to support partitioned table for text data source which seems  disabled in Spark 2.0. It seems the guide says it does not a JIRA only if it works as the same regardless of a PR.

**Comment by jurriaan** [comment]
I'll create a JIRA just to be sure, thanks!

**Comment by HyukjinKwon** [comment]
Let me cc @liancheng

**Comment by jurriaan** [comment]
@liancheng What do you think?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/text/TextSuite.scala [hunk 1]
```diff
@@ -65,6 +65,14 @@ class TextSuite extends QueryTest with SharedSQLContext {
     }
   }
 
+  test("support for partitioned reading") {
+    val df = spark.read.format("text").load(testDir)
```
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/text/TextSuite.scala]: This semantic makes sense to me.  But can we generate the test files dynamically instead of checking them in? (thats what we do in all the other tests)

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/text/TextSuite.scala [hunk 2]
```diff
@@ -65,6 +65,14 @@ class TextSuite extends QueryTest with SharedSQLContext {
...
+    val data = df.filter("year = '2015'").select("value").collect()
+
+    assert(data(0) == Row("2015-test"))
+    assert(data.length == 1)
```
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/text/TextSuite.scala]: What happens if you use `spark.read.text(...)`?  Does it still fail.  In that case I think we should just skip partition discovery.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/text/TextSuite.scala (hunk 1 above)
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/text/TextSuite.scala]: this is probably ok because we still have some test cases that do this (otherwise we'd need to generate the data and then make sure the directory structure is valid first)

**Comment by rxin** [comment]
It'd be better to change the directory name from partitioned to text-partitioned?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/text/TextSuite.scala (hunk 2 above)
Comment (jurriaan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/text/TextSuite.scala]: There's a JIRA for that too: https://issues.apache.org/jira/browse/SPARK-14463

You think I should fix that in this PR too?

Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/text/TextSuite.scala]: it'd be great to fix (but that can be done as a separate pr though)

**Comment by jurriaan** [comment]
@rxin changed the directory name to text-partitioned

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/text/TextSuite.scala (hunk 2 above)
Comment (jurriaan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/text/TextSuite.scala]: I'm new to this codebase (even new to Scala). But I'm happy to give that a go in a separate PR :)

Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/text/TextSuite.scala]: sounds great!

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/text/TextSuite.scala [hunk 3]
```diff
@@ -110,6 +118,10 @@ class TextSuite extends QueryTest with SharedSQLContext {
     Thread.currentThread().getContextClassLoader.getResource("text-suite.txt").toString
   }
 
+  private def testDir: String = {
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/text/TextSuite.scala]: actually this is a little bit far from the test case. since it is only used once, why not move this into the test case?

Comment (jurriaan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/text/TextSuite.scala]: Will do!

**Comment by jurriaan** [comment]
@rxin Noticed that https://issues.apache.org/jira/browse/SPARK-14463 was quite easily fixable, and as it's related to this bug, I've done it in this PR as well.

**Comment by HyukjinKwon** [comment]
@jurriaan It might be nicer if the title is `[SPARK-15323][SPARK-14463][SQL] ...` if this fixes the issue as well (So that both can be closed as you might already know).

**Comment by jurriaan** [comment]
Looks like Test build #58759 failed randomly, right? Looks like it's all green then, including the fix for SPARK-14463 :)

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala [hunk 4]
```diff
@@ -457,7 +457,8 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
    */
   @scala.annotation.varargs
   def text(paths: String*): Dataset[String] = {
-    format("text").load(paths : _*).as[String](sparkSession.implicits.newStringEncoder)
+    format("text").load(paths : _*).select("value")
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: i'm actually not sure if this is the right fix, because this is silently dropping the partition values. I was thinking just not returning anything here ... but I guess this is OK. We need to document it properly though.

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: Yeah, I think the semantic of dropping the columns is okay (I'm not sure the other option makes sense to me).  I'm a little concerned that we are still doing partition discovery, but I guess removing that is only an optimization.

**Comment by rxin** [comment]
Thanks - going to merge in master/2.0. I will submit a pr to add some documentation.
